src/predictor/gpu_predictor.cu

```diff
@@ -213,9 +213,10 @@ __global__ void PredictKernel(Data data,
 
 class DeviceModel {
  public:
-  dh::device_vector<RegTree::Node> nodes;
-  dh::device_vector<size_t> tree_segments;
-  dh::device_vector<int> tree_group;
+  // Need to lazily construct the vectors because GPU id is only known at runtime
+  std::unique_ptr<dh::device_vector<RegTree::Node>> nodes;
```

Comment:
I thought we are constructing the predictor lazily?

Reply:
The device vectors were being constructed before `cudaSetDevice()` was called. The device vectors need access to correct CUDA context at the time of construction, so I've delayed construction of the device vectors.

Reply:
This is a bit worrying if this is necessary, not sure if we can guarantee this behaviour across xgboost. Also wouldn't it be easier to place DeviceModel inside a unique pointer?

Reply:
Previously, this was not necessary, since Thrust's device vector built its memory resource (MR) lazily for each GPU it was being used. On the other hand, if we use RMM allocator with device vectors, then the correct CUDA context needs to be set (with `cudaSetDevice()`) prior to the construction of the device vector.

For now this line works, but in the longer term I can design a new device vector class that lazily constructs the device MR.

> Also wouldn't it be easier to place DeviceModel inside a unique pointer?

That won't work, because DeviceModel has a separate Init() function, and the correct CUDA context isn't set until we call Init() function.

https://github.com/dmlc/xgboost/blob/1d22a9be1cdeb53dfa9322c92541bc50e82f3c43/src/predictor/gpu_predictor.cu#L244-L245

Reply:
Is there an earlier point in the program that we can set the device? e.g. in the learner as soon as it receives the parameter gpu_id?

Reply:
Or you can use `HostDeviceVector`, which is lazy.

Reply:
Done. I replaced it with `HostDeviceVector`.